anival-github/AWS-Cloud-Developer-BE
# task-10

# Task 10 (Backend For Frontend)

[Link](https://github.com/rolling-scopes-school/aws/blob/main/aws-developer/10_backend_for_frontend/task.md) to the task
[Link](https://github.com/anival-github/nodejs-aws-shop-react/pull/6 ) to FE PR
[Link](https://github.com/anival-github/cart-api/pull/5) to Cart Api PR
[Link](http://d1ma6c5k4o6jd9.cloudfront.net/) to the front end deployed

Product Service API - direct [URL](https://8nzred66a6.execute-api.us-east-1.amazonaws.com/prod/)
Product Service API - list of products - direct [URL](https://8nzred66a6.execute-api.us-east-1.amazonaws.com/prod/products)

Cart Service API - direct [URL](http://production.eba-sjic3pmv.us-east-1.elasticbeanstalk.com) 
Cart Service API - retrieve cart items - direct [URL](http://production.eba-sjic3pmv.us-east-1.elasticbeanstalk.com/api/profile/cart)

Localhost BFF Service API [URL](http://localhost:3001)
Localhost BFF Service API - list of products, proxy [URL](http://localhost:3001) (redirect to deployed Product Service API)
Localhost BFF Service API - cart proxy [URL](http://localhost:3001/profile/cart) (redirect to deployed Cart Service API)

Deployed BFF Service API [URL](http://anival-github-bff-api-dev.us-east-1.elasticbeanstalk.com/) 
Deployed BFF Service API - list of products, proxy [URL](http://anival-github-bff-api-dev.us-east-1.elasticbeanstalk.com/products)
Deployed BFF Service API - cart proxy [URL](http://anival-github-bff-api-dev.us-east-1.elasticbeanstalk.com/profile/cart)

To test caching mechanism: 
URL to retrieve list of products: http://anival-github-bff-api-dev.us-east-1.elasticbeanstalk.com/products
cURL to create a product, you can modify JSON data in it: 
```
curl --location --request POST 'http://anival-github-bff-api-dev.us-east-1.elasticbeanstalk.com/products' \
--header 'Content-Type: application/json' \
--data-raw '{ "count": 2, "price": 100, "title": "aaaa1111", "description": "aaaa111", "image": "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTgWIECe8iudOKz8XuHDqzzY_PmXl74S7ew7w&usqp=CAU"}'
```

To test caching mechanism locally: 
URL to retrieve list of products: http://localhost:3001/products
cURL to create a product, you can modify JSON data in it: 
```
curl --location 'http://localhost:3001/products' \
--header 'Content-Type: application/json' \
--data '{ "count": 2, "price": 100, "title": "aaaa", "description": "aaaa", "image": "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTgWIECe8iudOKz8XuHDqzzY_PmXl74S7ew7w&usqp=CAU" }'
```

To pass authorisation, you need to add the following item into Front End local storage: 
```
localStorage.setItem('authorization_token', 'YW5pdmFsLWdpdGh1YjpURVNUX1BBU1NXT1JE')
```

Self-evaluation: 100/100

## Tasks

### Task 10.1

- [x] 1. Create a new service called `bff-service` at the same level as Product Service and Import Service.
- [x] 2. Create an application in this folder, that listens for all requests and redirects those requests to the appropriate services based on variables provided by the `.env` file.
- [x] 3. Workflow example that BFF Service should support:

- Make requests to BFF Service with URL in the following format: `{bff-service-url}/{recipient-service-name}?var1=someValue` 

    - `{bff-service-url}` - for example, http://localhost:3000
    - `{recipient-service-name}` - "cart" or "product" (you can use any other mapping of your choice)
    - `?var1=someValue` - query string

- Get `recipientURL` from the env variables using `{recipient-service-name}` as a key
- Get request `method` (GET, POST, etc.)
- Make a new request to the needed service using the appropriate `method` and `recipientURL`
- BFF Service should return the result of the recipient’s request

- [x] 4. If BFF Service cannot find `recipientURL` by the `{recipient-service-name}`, return a "Cannot process request" error message with status 502.
- [x] 5. BFF Service should return the same status code and error message that the recipient service returns to the BFF Service in case of any error on the recipient service side.

### Task 10.2

- [x] 1. Deploy BFF Service with Elastic Beanstalk.

- Application name must follow the following convention `{yours_github_account_login}-bff-api`
- Use the `--cname` option `{yours_github_account_login}-bff-api-{environment_name}`
- Use the `--single` option

- [x] 2. BFF Service should work only with requests from the Product Service and Cart Service.
- [x] 3. All Product Service and Cart Service methods should work correctly if requested via BFF Service

### Task 10.3

- [x] 1. Commit all your work to separate branch (e.g. `task-10` from the latest `master`) in your own repository.
- [x] 2. Create a pull request to the `master` branch.
- [x] 3. Submit link to the pull request to Crosscheck page in [RS App](https://app.rs.school).

## Evaluation criteria (80 points for covering all criteria)


## Application Functionality (**MUST HAVE**)

- [x] 1. Products representation on Home page should be based on Product Service API.
- [x] 2. Products are coming from Product DB.
- [x] 3. Product images are not randomly generated on client side. Product image, same as another product model information should be stored on BE side in Product DB.
- [x] 4. Products might be created through CSV product file import from client side.
- [x] 5. Cart might be created with appropriate product set.
- [x] 6. Auth logic should be in place

## Additional (optional) tasks

  How to test:
  - Get products list
  - Create new product
  - Get products list - result shouldn’t have new product
  - Wait more than 2 minutes
  - Get products list - result should have new product
